File: PassedProblems/257-binary-tree-paths.cpp
/*
 * 257-binary-tree-paths.cpp
 *
 *  Created on: Oct 4, 2015
 *      Author: hanzhe
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <iostream>
#include <sstream>
using namespace std;
#include <vector>
struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> re;
        if(root == NULL)
        	return re;
        vector<int> tPath;
        find(re, root, tPath);
        return re;
    }

    void find(vector<string> & re, TreeNode * r, vector<int>& path){
    	if(r == NULL)
    		return;
    	path.push_back(r->val);
    	if(r->left == NULL && r->right == NULL){
    		ostringstream os;
    		os << path[0];
    		for(int i = 1 ; i < path.size(); i ++)
    			os << "->" << path[i];
    		re.push_back(os.str());
    	}
    	find(re, r->left, path);
    	find(re, r->right, path);
    	path.pop_back();
    }
};

File: leetcode/python/257-binary-tree-paths.py
#!/usr/bin/env python
# encoding: utf-8


"""
Given a binary tree, return all root-to-leaf paths.

For example, given the following binary tree:

   1
 /   \
2     3
 \
  5
All root-to-leaf paths are:

["1->2->5", "1->3"]

"""


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# 3次过
class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        re = []
        def fn(node, ls):
            if node == None: return
            ls.append(node.val)
            if node.left == None and node.right == None:
                re.append(ls[:])
                ls.pop()
                return
            fn(node.left, ls)
            fn(node.right, ls)
            ls.pop()
        fn(root, [])
        re2 = []
        for x in re:
            re2.append("->".join(str(t) for t in x))
        return re2


